File: backend-v5/app/schemas/corporate.py
"""Pydantic schemas for Corporate Intelligence API."""
from datetime import date, datetime
from typing import Optional, List, Dict

from pydantic import BaseModel, Field


# ============================================================
# IRD Enrichment (embedded in company responses)
# ============================================================

class IRDEnrichmentSummary(BaseModel):
    """Lightweight IRD enrichment info attached to company records."""
    pan: str
    taxpayer_name_en: Optional[str] = None
    taxpayer_name_np: Optional[str] = None
    account_type: Optional[str] = None
    account_status: Optional[str] = None
    registration_date_bs: Optional[str] = None
    tax_office: Optional[str] = None
    is_personal: Optional[str] = None
    ward_no: Optional[str] = None
    vdc_municipality: Optional[str] = None
    phone_hash: Optional[str] = None
    mobile_hash: Optional[str] = None
    latest_tax_clearance_fy: Optional[str] = None
    tax_clearance_verified: Optional[bool] = None
    fetched_at: Optional[datetime] = None


# ============================================================
# Company with enrichment
# ============================================================

class CorporateCompanyResponse(BaseModel):
    """Company record with IRD enrichment data attached."""
    id: str
    external_id: str
    registration_number: int
    name_nepali: Optional[str] = None
    name_english: str
    registration_date_bs: Optional[str] = None
    registration_date_ad: Optional[date] = None
    company_type: Optional[str] = None
    company_type_category: Optional[str] = None
    company_address: Optional[str] = None
    district: Optional[str] = None
    province: Optional[str] = None
    last_communication_bs: Optional[str] = None
    pan: Optional[str] = None
    camis_company_id: Optional[int] = None
    cro_company_id: Optional[str] = None
    camis_enriched: bool = False
    camis_enriched_at: Optional[datetime] = None
    ird_enriched: bool = False
    ird_enriched_at: Optional[datetime] = None
    fetched_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # Attached IRD enrichment (None if not available)
    ird_enrichment: Optional[IRDEnrichmentSummary] = None
    # Convenience fields for list views
    ird_status: Optional[str] = None
    ird_taxpayer_name: Optional[str] = None
    director_count: int = 0
    linked_company_count: int = 0
    govt_contract_count: int = 0
    govt_contract_total_npr: Optional[float] = None


class DirectorResponse(BaseModel):
    """Director record for corporate intelligence views."""
    id: str
    company_id: Optional[str] = None
    name_en: str
    name_np: Optional[str] = None
    role: Optional[str] = None
    company_name_hint: Optional[str] = None
    source: str
    confidence: float = 1.0
    pan: Optional[str] = None
    citizenship_no: Optional[str] = None
    appointed_date: Optional[date] = None
    resigned_date: Optional[date] = None


# ============================================================
# PAN Investigation
# ============================================================

class RiskFlag(BaseModel):
    """A single risk indicator."""
    severity: str = Field(description="HIGH, MEDIUM, or LOW")
    category: str = Field(description="e.g. tax_compliance, address_clustering, pan_sharing")
    description: str
    details: Optional[Dict] = None


class PANInvestigationResponse(BaseModel):
    """Full PAN investigation result."""
    pan: str
    companies: List[CorporateCompanyResponse]
    ird: Optional[IRDEnrichmentSummary] = None
    risk_flags: List[RiskFlag] = Field(default_factory=list)


# ============================================================
# Search / List
# ============================================================

class CorporateSearchResponse(BaseModel):
    """Paginated company search results."""
    items: List[CorporateCompanyResponse]
    total: int
    page: int = 1
    limit: int = 50
    has_more: bool = False


# ============================================================
# Company Detail
# ============================================================

class CompanyDetailResponse(CorporateCompanyResponse):
    """Full company profile with directors and IRD data (flattened)."""
    ird_status: Optional[str] = None
    ird_taxpayer_name: Optional[str] = None
    directors: List[DirectorResponse] = Field(default_factory=list)
    ird: Optional[IRDEnrichmentSummary] = None
    risk_flags: List[RiskFlag] = Field(default_factory=list)
    govt_procurement_summary: Optional["GovtProcurementSummary"] = None


class GovtProcurementEntitySummary(BaseModel):
    """Aggregate procurement totals per government procuring entity."""
    name: str
    contract_count: int
    total_value_npr: float


class GovtProcurementContract(BaseModel):
    """Individual government contract record linked to a company."""
    id: str
    contractor_name: Optional[str] = None
    procuring_entity: Optional[str] = None
    ifb_number: Optional[str] = None
    project_name: Optional[str] = None
    procurement_type: Optional[str] = None
    contract_award_date: Optional[str] = None
    contract_amount_npr: Optional[float] = None
    fiscal_year_bs: Optional[str] = None
    district: Optional[str] = None
    source_url: Optional[str] = None


class GovtProcurementSummary(BaseModel):
    """Expanded procurement context for company detail drawer."""
    linked_contractor_names: List[str] = Field(default_factory=list)
    procuring_entities: List[GovtProcurementEntitySummary] = Field(default_factory=list)
    contracts: List[GovtProcurementContract] = Field(default_factory=list)


# ============================================================
# Dashboard Stats
# ============================================================

class CorporateStatsResponse(BaseModel):
    """Dashboard-level corporate intelligence stats."""
    total_companies: int
    companies_with_pan: int
    pan_coverage_pct: float = Field(description="Percentage of companies with PAN")
    camis_enriched_count: int = 0
    ird_enriched_count: int
    ird_enrichment_pct: float = Field(description="Percentage of PAN-holding companies with IRD data")
    total_directors: int
    companies_by_type: Dict[str, int] = Field(default_factory=dict)
    companies_by_province: Dict[str, int] = Field(default_factory=dict)
    top_districts: Dict[str, int] = Field(default_factory=dict)
    risk_summary: Dict[str, int] = Field(default_factory=dict, description="Count of risk flags by severity")


# ============================================================
# Shared Directors (Network)
# ============================================================

class PhoneLinkedCompany(BaseModel):
    """A company linked via shared phone/mobile number hash."""
    company_id: str
    company_name: str
    pan: Optional[str] = None
    district: Optional[str] = None
    company_address: Optional[str] = None
    ird_status: Optional[str] = None
    match_type: str = Field(description="phone, mobile, or both")


class PhoneLinksResponse(BaseModel):
    """Companies linked by shared phone/mobile number."""
    company_id: str
    company_name: str
    links: List[PhoneLinkedCompany] = Field(default_factory=list)


# ============================================================
# Phone Clusters (global view)
# ============================================================

class PhoneClusterCompany(BaseModel):
    """A company within a phone/mobile hash cluster."""
    company_id: str
    company_name: str
    pan: Optional[str] = None
    registration_number: Optional[int] = None
    district: Optional[str] = None
    company_address: Optional[str] = None
    ird_status: Optional[str] = None


class PhoneCluster(BaseModel):
    """A cluster of companies sharing the same phone or mobile hash."""
    cluster_id: str = Field(description="Stable cluster identifier for analyst graphing")
    hash_type: str = Field(description="phone, mobile, or both")
    company_count: int
    first_registered: PhoneClusterCompany
    companies: List[PhoneClusterCompany] = Field(
        default_factory=list,
        description="Cluster members (may be truncated for performance)",
    )


class PhoneClustersResponse(BaseModel):
    """All phone/mobile hash clusters across the dataset."""
    clusters: List[PhoneCluster]
    total_clusters: int
    total_linked_companies: int


# ============================================================
# Analyst Cluster Groups (manual analyst graphing)
# ============================================================

class AnalystClusterNode(BaseModel):
    """A phone cluster node saved into an analyst-defined group graph."""
    cluster_id: str = Field(min_length=1, max_length=160)
    label: str = Field(min_length=1, max_length=300)
    hash_type: Optional[str] = None
    company_count: Optional[int] = None
    first_registered_company_id: Optional[str] = None
    first_registered_company_name: Optional[str] = None


class AnalystClusterEdge(BaseModel):
    """A manual relationship between two saved phone cluster nodes."""
    id: Optional[str] = Field(default=None, max_length=120)
    source_cluster_id: str = Field(min_length=1, max_length=160)
    target_cluster_id: str = Field(min_length=1, max_length=160)
    label: str = Field(min_length=1, max_length=120)
    bidirectional: bool = False


class AnalystClusterGroupCreate(BaseModel):
    """Create payload for analyst-defined cluster group graph."""
    name: str = Field(min_length=1, max_length=255)
    description: Optional[str] = None
    main_cluster_id: Optional[str] = None
    clusters: List[AnalystClusterNode] = Field(default_factory=list)
    edges: List[AnalystClusterEdge] = Field(default_factory=list)


class AnalystClusterGroupUpdate(BaseModel):
    """Patch payload for analyst-defined cluster group graph."""
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    description: Optional[str] = None
    main_cluster_id: Optional[str] = None
    clusters: Optional[List[AnalystClusterNode]] = None
    edges: Optional[List[AnalystClusterEdge]] = None


class AnalystClusterGroupResponse(BaseModel):
    """Stored analyst-defined cluster group graph."""
    id: str
    name: str
    description: Optional[str] = None
    main_cluster_id: Optional[str] = None
    clusters: List[AnalystClusterNode] = Field(default_factory=list)
    edges: List[AnalystClusterEdge] = Field(default_factory=list)
    created_by_id: Optional[str] = None
    created_by_name: Optional[str] = None
    updated_by_id: Optional[str] = None
    updated_by_name: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class AnalystClusterGroupListResponse(BaseModel):
    """List response for analyst cluster groups."""
    items: List[AnalystClusterGroupResponse]
    total: int


class SharedDirectorLink(BaseModel):
    """A director who appears in multiple companies."""
    director_name: str
    director_role: Optional[str] = None
    source: str
    linked_company_id: str
    linked_company_name: str
    linked_company_pan: Optional[str] = None


class SharedDirectorsResponse(BaseModel):
    """Companies sharing directors with a given company."""
    company_id: str
    company_name: str
    shared_links: List[SharedDirectorLink] = Field(default_factory=list)
    unique_linked_companies: int = 0


# ============================================================
# Address Clusters
# ============================================================

class AddressCluster(BaseModel):
    """An address with many registered companies (shell company signal)."""
    address: str
    company_count: int
    companies: List[CorporateCompanyResponse] = Field(default_factory=list)
    district: Optional[str] = None
    province: Optional[str] = None


class AddressClustersResponse(BaseModel):
    """List of address clusters."""
    clusters: List[AddressCluster]
    total_clusters: int


# ============================================================
# Risk Flags
# ============================================================

class CompanyRiskEntry(BaseModel):
    """A company with its risk flags."""
    company: CorporateCompanyResponse
    risk_flags: List[RiskFlag] = Field(default_factory=list)


class RiskFlagsResponse(BaseModel):
    """Paginated risk flags across companies."""
    items: List[CompanyRiskEntry]
    total: int
    page: int = 1
    limit: int = 50


# ============================================================
# Advanced Analytics: Beneficial Ownership
# ============================================================

class BeneficialOwnerCompany(BaseModel):
    """A company linked to a beneficial owner."""
    id: str
    name_english: str
    pan: Optional[str] = None
    district: Optional[str] = None
    role: Optional[str] = None


class BeneficialOwner(BaseModel):
    """A person who is a director of multiple companies."""
    name: str
    citizenship_no: Optional[str] = None
    total_companies: int
    companies: List[BeneficialOwnerCompany] = Field(default_factory=list)
    match_type: str = Field(description="citizenship_no or name_en")


class BeneficialOwnersResponse(BaseModel):
    """Beneficial ownership discovery results."""
    owners: List[BeneficialOwner]
    total: int


# ============================================================
# Advanced Analytics: Shell Company Scoring
# ============================================================

class ShellCompanyScore(BaseModel):
    """A company with its shell company risk score."""
    id: str
    name_english: str
    pan: Optional[str] = None
    company_address: Optional[str] = None
    district: Optional[str] = None
    registration_date_ad: Optional[str] = None
    score: int
    factors: List[str] = Field(default_factory=list)


class ShellCompanyScoresResponse(BaseModel):
    """Shell company scoring results."""
    companies: List[ShellCompanyScore]
    total_scored: int


# ============================================================
# Advanced Analytics: Tax Compliance
# ============================================================

class DistrictCompliance(BaseModel):
    """Tax compliance stats for a district."""
    district: str
    total: int
    nonfiler_count: int


class TypeCompliance(BaseModel):
    """Tax compliance stats for a company type category."""
    category: str
    total: int
    nonfiler_count: int


class TaxComplianceStatsResponse(BaseModel):
    """Tax compliance dashboard statistics."""
    total_pans: int
    active_filers: int
    non_filers: int
    cancelled: int
    unknown: int
    status_breakdown: Dict[str, int] = Field(default_factory=dict)
    by_district: List[DistrictCompliance] = Field(default_factory=list)
    by_company_type: List[TypeCompliance] = Field(default_factory=list)


# ============================================================
# Advanced Analytics: Network Stats
# ============================================================

class TopDirector(BaseModel):
    """A director who serves on many boards."""
    name: str
    citizenship_no: Optional[str] = None
    company_count: int


class TopAddress(BaseModel):
    """An address with many registered companies."""
    address: str
    district: Optional[str] = None
    company_count: int


class PANSharingGroup(BaseModel):
    """A group of companies sharing the same PAN."""
    pan: str
    company_count: int
    company_names: Optional[str] = None


class NetworkSummary(BaseModel):
    """Summary counts for network statistics."""
    total_unique_directors: int
    multi_board_directors: int
    total_pan_sharing_groups: int


class NetworkStatsResponse(BaseModel):
    """Corporate network statistics."""
    top_directors: List[TopDirector] = Field(default_factory=list)
    top_addresses: List[TopAddress] = Field(default_factory=list)
    pan_sharing_groups: List[PANSharingGroup] = Field(default_factory=list)
    summary: NetworkSummary


# ============================================================
# Advanced Analytics: Registration Patterns
# ============================================================

class YearlyRegistration(BaseModel):
    """Registration count for a year."""
    year: int
    count: int


class MonthlyRegistration(BaseModel):
    """Registration count for a month."""
    year: int
    month: int
    count: int


class PeakDate(BaseModel):
    """A peak registration date."""
    date: Optional[str] = None
    count: int


class SameDayCluster(BaseModel):
    """A cluster of companies registered at the same address on the same day."""
    address: str
    date: Optional[str] = None
    count: int


class RegistrationPatternsResponse(BaseModel):
    """Registration pattern analysis results."""
    yearly: List[YearlyRegistration] = Field(default_factory=list)
    monthly: List[MonthlyRegistration] = Field(default_factory=list)
    peak_dates: List[PeakDate] = Field(default_factory=list)
    same_day_clusters: List[SameDayCluster] = Field(default_factory=list)
    anomaly_threshold: float = 0.0

File: frontend/src/api/corporate.ts
/**
 * Corporate Intelligence API client
 * V6: Company Explorer, PAN Investigation, Director Networks
 */
import { apiClient } from './client'

// ── Types ──────────────────────────────────────────────────────

export interface CompanyRecord {
  id: string
  registration_number: number
  name_english: string
  name_nepali: string | null
  registration_date_bs: string | null
  registration_date_ad: string | null
  company_type: string | null
  company_type_category: string | null
  company_address: string | null
  district: string | null
  province: string | null
  pan: string | null
  camis_enriched: boolean
  ird_enriched: boolean
  ird_status: string | null
  ird_taxpayer_name: string | null
  director_count?: number
  linked_company_count?: number
  govt_contract_count?: number
  govt_contract_total_npr?: number | null
}

export interface CompanyDetail extends CompanyRecord {
  camis_company_id: number | null
  cro_company_id: string | null
  camis_enriched_at: string | null
  ird_enriched_at: string | null
  last_communication_bs: string | null
  directors: DirectorRecord[]
  ird: IRDRecord | null
  govt_procurement_summary?: GovtProcurementSummary | null
}

export interface GovtProcurementEntitySummary {
  name: string
  contract_count: number
  total_value_npr: number
}

export interface GovtProcurementContract {
  id: string
  contractor_name: string | null
  procuring_entity: string | null
  ifb_number: string | null
  project_name: string | null
  procurement_type: string | null
  contract_award_date: string | null
  contract_amount_npr: number | null
  fiscal_year_bs: string | null
  district: string | null
  source_url: string | null
}

export interface GovtProcurementSummary {
  linked_contractor_names: string[]
  procuring_entities: GovtProcurementEntitySummary[]
  contracts: GovtProcurementContract[]
}

export interface DirectorRecord {
  id: string
  name_en: string
  name_np: string | null
  role: string | null
  source: string
  confidence: number
  pan: string | null
  citizenship_no: string | null
  appointed_date: string | null
  resigned_date: string | null
  company_id: string | null
  company_name: string | null
}

export interface IRDRecord {
  pan: string
  taxpayer_name_en: string | null
  taxpayer_name_np: string | null
  account_type: string | null
  account_status: string | null
  registration_date_bs: string | null
  tax_office: string | null
  filing_period: string | null
  is_personal: string | null
  ward_no: string | null
  vdc_municipality: string | null
  latest_tax_clearance_fy: string | null
  tax_clearance_verified: boolean | null
}

export interface RiskFlag {
  company_id: string
  company_name: string
  pan: string | null
  flag_type: string
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW'
  description: string
}

export interface PhoneLinkedCompany {
  company_id: string
  company_name: string
  pan: string | null
  district: string | null
  company_address: string | null
  ird_status: string | null
  match_type: string
}

export interface PhoneLinksResponse {
  company_id: string
  company_name: string
  links: PhoneLinkedCompany[]
}

// ── Phone Clusters (global view) ──────────────────────────────

export interface PhoneClusterCompany {
  company_id: string
  company_name: string
  pan: string | null
  registration_number: number | null
  district: string | null
  company_address: string | null
  ird_status: string | null
}

export interface PhoneCluster {
  cluster_id: string
  hash_type: 'phone' | 'mobile' | 'both'
  company_count: number
  first_registered: PhoneClusterCompany
  companies?: PhoneClusterCompany[]
}

export interface PhoneClustersResponse {
  clusters: PhoneCluster[]
  total_clusters: number
  total_linked_companies: number
}

export interface AnalystClusterNode {
  cluster_id: string
  label: string
  hash_type?: string | null
  company_count?: number | null
  first_registered_company_id?: string | null
  first_registered_company_name?: string | null
}

export interface AnalystClusterEdge {
  id?: string | null
  source_cluster_id: string
  target_cluster_id: string
  label: string
  bidirectional: boolean
}

export interface AnalystClusterGroup {
  id: string
  name: string
  description: string | null
  main_cluster_id: string | null
  clusters: AnalystClusterNode[]
  edges: AnalystClusterEdge[]
  created_by_id: string | null
  created_by_name: string | null
  updated_by_id: string | null
  updated_by_name: string | null
  created_at: string | null
  updated_at: string | null
}

export interface AnalystClusterGroupListResponse {
  items: AnalystClusterGroup[]
  total: number
}

export interface AnalystClusterGroupCreatePayload {
  name: string
  description?: string | null
  main_cluster_id?: string | null
  clusters?: AnalystClusterNode[]
  edges?: AnalystClusterEdge[]
}

export interface AnalystClusterGroupUpdatePayload {
  name?: string
  description?: string | null
  main_cluster_id?: string | null
  clusters?: AnalystClusterNode[]
  edges?: AnalystClusterEdge[]
}

export interface PANInvestigation {
  pan: string
  companies: CompanyRecord[]
  ird: IRDRecord | null
  directors: DirectorRecord[]
  risk_flags: RiskFlag[]
}

export interface CorporateStats {
  total_companies: number
  companies_with_pan: number
  pan_coverage_pct: number
  camis_enriched_count: number
  ird_enriched_count: number
  ird_enrichment_pct: number
  total_directors: number
  companies_by_type: Record<string, number>
  companies_by_province: Record<string, number>
  top_districts: Record<string, number>
  risk_summary: Record<string, number>
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  limit: number
  pages: number
  has_more?: boolean
}

export interface CompanySearchParams {
  q?: string
  district?: string
  company_type_category?: string
  has_pan?: boolean
  ird_status?: string
  has_cluster?: boolean
  sort?: string
  sort_dir?: 'asc' | 'desc'
  page?: number
  limit?: number
}

// ── API Functions ──────────────────────────────────────────────

export async function getCorporateStats(): Promise<CorporateStats> {
  const { data } = await apiClient.get('/corporate/stats')
  return data
}

export async function searchCompanies(params: CompanySearchParams): Promise<PaginatedResponse<CompanyRecord>> {
  const { data } = await apiClient.get('/corporate/companies', { params })
  const total = Number(data?.total ?? 0)
  const limit = Number(data?.limit ?? params.limit ?? 50)
  return {
    ...data,
    total,
    limit,
    pages: Math.max(1, Math.ceil(total / Math.max(1, limit))),
  } as PaginatedResponse<CompanyRecord>
}

export async function getCompanyDetail(id: string): Promise<CompanyDetail> {
  const { data } = await apiClient.get(`/corporate/companies/${id}`)
  // Backend returns flat company fields + ird + directors + risk_flags
  return data as CompanyDetail
}

export async function investigatePAN(pan: string): Promise<PANInvestigation> {
  const { data } = await apiClient.get(`/corporate/pan/${pan}`)
  return data
}

export async function getPhoneLinks(companyId: string): Promise<PhoneLinksResponse> {
  const { data } = await apiClient.get(`/corporate/companies/${companyId}/phone-links`)
  return data
}

export async function getSharedDirectors(companyId: string): Promise<DirectorRecord[]> {
  const { data } = await apiClient.get(`/corporate/shared-directors/${companyId}`)
  if (Array.isArray(data)) return data as DirectorRecord[]
  const sharedLinks = Array.isArray(data?.shared_links) ? data.shared_links : []
  return sharedLinks.map((link: any, index: number) => ({
    id: `${companyId}-${link.linked_company_id}-${index}`,
    name_en: link.director_name ?? 'Unknown',
    name_np: null,
    role: link.director_role ?? null,
    source: link.source ?? 'corporate_shared_director',
    confidence: 1,
    pan: link.linked_company_pan ?? null,
    citizenship_no: null,
    appointed_date: null,
    resigned_date: null,
    company_id: link.linked_company_id ?? null,
    company_name: link.linked_company_name ?? null,
  }))
}

export async function getAddressClusters(minCompanies: number = 5): Promise<Array<{ address: string; count: number; companies: CompanyRecord[] }>> {
  const { data } = await apiClient.get('/corporate/address-clusters', { params: { min_companies: minCompanies } })
  const clusters = (data?.clusters ?? data ?? []) as Array<{
    address: string
    count?: number
    company_count?: number
    companies: CompanyRecord[]
  }>
  return clusters.map((cluster) => ({
    address: cluster.address,
    count: cluster.count ?? cluster.company_count ?? 0,
    companies: cluster.companies ?? [],
  }))
}

export async function getPhoneClusters(params: {
  limit?: number
  min_companies?: number
  max_members_per_cluster?: number
} = {}): Promise<PhoneClustersResponse> {
  const firstAttemptParams = {
    limit: params.limit ?? 60,
    min_companies: params.min_companies ?? 2,
    max_members_per_cluster: params.max_members_per_cluster ?? 200,
  }

  try {
    const { data } = await apiClient.get('/corporate/phone-clusters', {
      params: firstAttemptParams,
      timeout: 90_000,
    })
    return data
  } catch (error: any) {
    // Timeout fallback: retry with a smaller payload.
    const fallbackParams = {
      limit: Math.min(firstAttemptParams.limit, 40),
      min_companies: firstAttemptParams.min_companies,
      max_members_per_cluster: Math.min(firstAttemptParams.max_members_per_cluster, 60),
    }
    const isTimeout = error?.code === 'ECONNABORTED' || /timeout/i.test(String(error?.message ?? ''))
    if (!isTimeout) throw error

    const { data } = await apiClient.get('/corporate/phone-clusters', {
      params: fallbackParams,
      timeout: 90_000,
    })
    return data
  }
}

export async function listAnalystClusterGroups(params: {
  only_mine?: boolean
  limit?: number
} = {}): Promise<AnalystClusterGroupListResponse> {
  const { data } = await apiClient.get('/corporate/cluster-groups', {
    params: {
      only_mine: params.only_mine ?? false,
      limit: params.limit ?? 200,
    },
  })
  return {
    items: data?.items ?? [],
    total: Number(data?.total ?? 0),
  }
}

export async function getAnalystClusterGroup(groupId: string): Promise<AnalystClusterGroup> {
  const { data } = await apiClient.get(`/corporate/cluster-groups/${groupId}`)
  return data as AnalystClusterGroup
}

export async function createAnalystClusterGroup(
  payload: AnalystClusterGroupCreatePayload
): Promise<AnalystClusterGroup> {
  const { data } = await apiClient.post('/corporate/cluster-groups', payload)
  return data as AnalystClusterGroup
}

export async function updateAnalystClusterGroup(
  groupId: string,
  payload: AnalystClusterGroupUpdatePayload
): Promise<AnalystClusterGroup> {
  const { data } = await apiClient.patch(`/corporate/cluster-groups/${groupId}`, payload)
  return data as AnalystClusterGroup
}

export async function deleteAnalystClusterGroup(groupId: string): Promise<void> {
  await apiClient.delete(`/corporate/cluster-groups/${groupId}`)
}

export async function getRiskFlags(minSeverity: string = 'MEDIUM'): Promise<RiskFlag[]> {
  const { data } = await apiClient.get('/corporate/risk-flags', { params: { min_severity: minSeverity } })
  const items = Array.isArray(data) ? data : (data?.items ?? [])
  return items.flatMap((entry: any) => {
    const company = entry?.company ?? {}
    const companyFlags = Array.isArray(entry?.risk_flags) ? entry.risk_flags : []
    return companyFlags.map((flag: any) => ({
      company_id: String(company.id ?? ''),
      company_name: company.name_english ?? 'Unknown company',
      pan: company.pan ?? null,
      flag_type: flag.category ?? 'unknown',
      severity: (flag.severity ?? 'LOW') as RiskFlag['severity'],
      description: flag.description ?? '',
    }))
  })
}


// ── Advanced Analytics Types ──────────────────────────────────

export interface BeneficialOwnerCompany {
  id: string
  name_english: string
  pan: string | null
  district: string | null
  role: string | null
}

export interface BeneficialOwner {
  name: string
  citizenship_no: string | null
  total_companies: number
  companies: BeneficialOwnerCompany[]
  match_type: string
}

export interface BeneficialOwnersResponse {
  owners: BeneficialOwner[]
  total: number
}

export interface ShellCompanyScore {
  id: string
  name_english: string
  pan: string | null
  company_address: string | null
  district: string | null
  registration_date_ad: string | null
  score: number
  factors: string[]
}

export interface ShellCompanyScoresResponse {
  companies: ShellCompanyScore[]
  total_scored: number
}

export interface DistrictCompliance {
  district: string
  total: number
  nonfiler_count: number
}

export interface TypeCompliance {
  category: string
  total: number
  nonfiler_count: number
}

export interface TaxComplianceStats {
  total_pans: number
  active_filers: number
  non_filers: number
  cancelled: number
  unknown: number
  status_breakdown: Record<string, number>
  by_district: DistrictCompliance[]
  by_company_type: TypeCompliance[]
}

export interface TopDirector {
  name: string
  citizenship_no: string | null
  company_count: number
}

export interface TopAddress {
  address: string
  district: string | null
  company_count: number
}

export interface PANSharingGroup {
  pan: string
  company_count: number
  company_names: string | null
}

export interface NetworkSummary {
  total_unique_directors: number
  multi_board_directors: number
  total_pan_sharing_groups: number
}

export interface NetworkStats {
  top_directors: TopDirector[]
  top_addresses: TopAddress[]
  pan_sharing_groups: PANSharingGroup[]
  summary: NetworkSummary
}

export interface YearlyRegistration {
  year: number
  count: number
}

export interface MonthlyRegistration {
  year: number
  month: number
  count: number
}

export interface PeakDate {
  date: string | null
  count: number
}

export interface SameDayCluster {
  address: string
  date: string | null
  count: number
}

export interface RegistrationPatterns {
  yearly: YearlyRegistration[]
  monthly: MonthlyRegistration[]
  peak_dates: PeakDate[]
  same_day_clusters: SameDayCluster[]
  anomaly_threshold: number
}


// ── Advanced Analytics API Functions ──────────────────────────

export async function getBeneficialOwners(
  minCompanies: number = 3,
  limit: number = 50,
): Promise<BeneficialOwnersResponse> {
  const { data } = await apiClient.get('/corporate/analytics/beneficial-owners', {
    params: { min_companies: minCompanies, limit },
  })
  return data
}

export async function getShellScores(limit: number = 100): Promise<ShellCompanyScoresResponse> {
  const { data } = await apiClient.get('/corporate/analytics/shell-scores', {
    params: { limit },
  })
  return data
}

export async function getTaxCompliance(): Promise<TaxComplianceStats> {
  const { data } = await apiClient.get('/corporate/analytics/tax-compliance')
  return data
}

export async function getNetworkStats(): Promise<NetworkStats> {
  const { data } = await apiClient.get('/corporate/analytics/network-stats')
  return data
}

export async function getRegistrationPatterns(): Promise<RegistrationPatterns> {
  const { data } = await apiClient.get('/corporate/analytics/registration-patterns')
  return data
}


// ── Timeline ────────────────────────────────────────────────

export interface TimelineBucket {
  period: string
  count: number
}

export interface TimelineResponse {
  items: TimelineBucket[]
  group_by: string
}

export interface TimelineEvent {
  id: string
  content: string
  start: string
  group: string
  type: string
  pan: string | null
}

export async function getRegistrationTimeline(params: {
  group_by?: 'month' | 'year'
  district?: string
  company_type?: string
} = {}): Promise<TimelineResponse> {
  const { data } = await apiClient.get('/corporate/timeline', { params })
  return data
}

export async function getRegistrationEvents(params: {
  start_date?: string
  end_date?: string
  district?: string
  limit?: number
} = {}): Promise<TimelineEvent[]> {
  const { data } = await apiClient.get('/corporate/timeline/events', { params })
  return data
}
